hack/generator/pkg/codegen/pipeline_create_arm_types.go

```diff
@@ -13,275 +13,118 @@ import (
-			}
-
-			armSpecDef, ok := kubeNameToArmDefs[specDef.Name()]
-			if !ok {
-				return nil, nil, errors.Errorf("couldn't find arm def matching kube def %q", specDef.Name())
-			}
-
-			modifiedSpec, err := addArmConversionInterface(specDef, armSpecDef, idFactory, armconversion.SpecType)
-			if err != nil {
-				return nil, nil, err
-			}
-
-			names := []astmodel.TypeName{modifiedSpec.Name()}
-			defs := []astmodel.TypeDefinition{modifiedSpec}
-
-			statusType := astmodel.IgnoringErrors(resourceType.StatusType())
-			if statusType != nil {
-				statusName, ok := statusType.(astmodel.TypeName)
-				if !ok {
-					return nil, nil, errors.Errorf("expected Status type to be a name, instead was %T", statusType)
-				}
-
-				statusDef, ok := definitions[statusName]
-				if !ok {
-					return nil, nil, errors.Errorf("could not find Status type %v", statusName)
-				}
-
-				armStatusDef, ok := kubeNameToArmDefs[statusName]
-				if !ok {
-					return nil, nil, errors.Errorf("could not find ARM Status type %v", statusName)
-				}
-
-				newStatus, err := addArmConversionInterface(statusDef, armStatusDef, idFactory, armconversion.StatusType)
-				if err != nil {
-					return nil, nil, err
-				}
-
-				names = append(names, newStatus.Name())
-				defs = append(defs, newStatus)
-			}
-
-			return names, defs, nil
-		},
-		// Other defs handler
-		func(def astmodel.TypeDefinition) (astmodel.TypeDefinition, error) {
-			armDef, ok := kubeNameToArmDefs[def.Name()]
-			if !ok {
-				return astmodel.TypeDefinition{}, errors.Errorf("couldn't find arm def matching kube def %q", def.Name())
-			}
-
-			modifiedDef, err := addArmConversionInterface(def, armDef, idFactory, armconversion.OrdinaryType)
-			if err != nil {
-				return astmodel.TypeDefinition{}, errors.Wrapf(err, "failed to add ARM conversion interface to %q", def.Name())
-			}
-
-			return modifiedDef, nil
-		})
+type armTypeCreator struct {
+	definitions astmodel.Types
+	idFactory   astmodel.IdentifierFactory
 }
 
-func iterDefs(
-	definitions astmodel.Types,
-	resourceHandler func(name astmodel.TypeName, resourceType *astmodel.ResourceType) ([]astmodel.TypeName, []astmodel.TypeDefinition, error),
-	otherDefsHandler func(def astmodel.TypeDefinition) (astmodel.TypeDefinition, error)) (astmodel.Types, error) {
-
-	newDefs := make(astmodel.Types)
-	actionedDefs := make(map[astmodel.TypeName]struct{})
-
-	// TODO: a better way to do all this following would be if we had Spec/Status tags
-	// for Spec/Status types. Then we would only do one pass over all types and
-	// just examine the flags, rather than walking all resources first and
-	// the remaining definitions in a different pass… and we wouldn’t have to deal with
-	// multiple uses of the same type (see the panic below), etc, etc…
-
-	// Do all the resources first - this ensures we avoid handling a spec before we've processed
-	// its associated resource.
+// TODO: Generic method on astmodel.Types?
```

Comment:
They're not going to be available that soon! ;)
How do you mean generic here? Is it just a matter of moving it somewhere shared?

Reply:
I just mean that this could be rephrased into a more official sounding name and moved to `astmodel.Types` as a built-in filter, as it seems to be an operation that we do quite frequently ("Find me the specs of all of the resources").

Sometimes though we need the spec and the resource (or the name of the resource), and sometimes we just need the spec, so we'd need to think about what we wanted the return type to be

Reply:
This TODO was just me attempting to ask the question: Do folks think that this operation is common enough for it to be promoted to a method someplace?